Terminal read improvements

MCUdude opened the pull request:
Addresses issue #1201.

This PR improves the terminal read functionality.

<details>
<summary>It's now possible to run <b>read</b> or <b>read [mem]</b>, and the command will continue where the previous left off</summary>

```
$ ./avrdude -pavr128db48 -cpkobn_updi -t -B0.5

         Vtarget                      : 3.31 V
         PDI/UPDI clock Xmega/megaAVR : 2000 kHz
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e970c (probably avr128db48)
avrdude> read
>>> read 
Usage: read <memory> <addr> <len>
       read <memory> <addr> ...
       read <memory> <addr>
       read <memory> ...
       read <memory>
avrdude> read flash 0 0x20
>>> read flash 0 0x20 

Reading | ################################################## | 100% 0.03 s 

00000  47 c0 00 00 55 c0 00 00  53 c0 00 00 51 c0 00 00  |G...U...S...Q...|
00010  4f c0 00 00 4d c0 00 00  4b c0 00 00 49 c0 00 00  |O...M...K...I...|

avrdude> read flash
>>> read flash 

Reading | ################################################## | 100% 0.00 s 

00020  47 c0 00 00 45 c0 00 00  43 c0 00 00 41 c0 00 00  |G...E...C...A...|
00030  3f c0 00 00 3d c0 00 00  3b c0 00 00 39 c0 00 00  |?...=...;...9...|

avrdude> read
>>> read 

Reading | ################################################## | 100% 0.00 s 

00040  37 c0 00 00 35 c0 00 00  33 c0 00 00 31 c0 00 00  |7...5...3...1...|
00050  2f c0 00 00 2d c0 00 00  2b c0 00 00 29 c0 00 00  |/...-...+...)...|

avrdude> read eeprom 0 0x10
>>> read eeprom 0 0x10 

Reading | ################################################## | 100% 0.06 s 

0000  48 65 6c 6c 6f 20 57 6f  72 6c 64 20 45 45 50 52  |Hello World EEPR|

avrdude> read
>>> read 

Reading | ################################################## | 100% 0.06 s 

0010  4f 4d 20 77 72 69 74 65  20 74 65 73 74 00 ff ff  |OM write test...|

avrdude> read flash
>>> read flash 

Reading | ################################################## | 100% 0.00 s 

00060  27 c0 00 00 97 c0 00 00  23 c0 00 00 21 c0 00 00  |'.......#...!...|
00070  1f c0 00 00 1d c0 00 00  1b c0 00 00 19 c0 00 00  |................|

avrdude> read
>>> read 

Reading | ################################################## | 100% 0.00 s 

00080  17 c0 00 00 15 c0 00 00  13 c0 00 00 11 c0 00 00  |................|
00090  11 24 1f be cf ef cd bf  df e3 de bf 28 e2 a0 e0  |.$..........(...|

avrdude> quit
>>> quit 
avrdude> 
avrdude done.  Thank you.
```

</details>


<details>
<summary><b>read/dump</b> now support memory rollover</summary>

```
$ ./avrdude -pavr128db48 -cpkobn_updi -t -B0.5

         Vtarget                      : 3.31 V
         PDI/UPDI clock Xmega/megaAVR : 2000 kHz
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e970c (probably avr128db48)
avrdude> read flash 0x1ffd3 0x100
>>> read flash 0x1ffd3 0x100 

Reading | ################################################## | 100% 0.06 s 

1ffd3  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
1ffe3  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
1fff3  ff ff ff ff ff ff ff ff  ff ff ff ff ff           |.............   |
00000  47 c0 00 00 55 c0 00 00  53 c0 00 00 51 c0 00 00  |G...U...S...Q...|
00010  4f c0 00 00 4d c0 00 00  4b c0 00 00 49 c0 00 00  |O...M...K...I...|
00020  47 c0 00 00 45 c0 00 00  43 c0 00 00 41 c0 00 00  |G...E...C...A...|
00030  3f c0 00 00 3d c0 00 00  3b c0 00 00 39 c0 00 00  |?...=...;...9...|
00040  37 c0 00 00 35 c0 00 00  33 c0 00 00 31 c0 00 00  |7...5...3...1...|
00050  2f c0 00 00 2d c0 00 00  2b c0 00 00 29 c0 00 00  |/...-...+...)...|
00060  27 c0 00 00 97 c0 00 00  23 c0 00 00 21 c0 00 00  |'.......#...!...|
00070  1f c0 00 00 1d c0 00 00  1b c0 00 00 19 c0 00 00  |................|
00080  17 c0 00 00 15 c0 00 00  13 c0 00 00 11 c0 00 00  |................|
00090  11 24 1f be cf ef cd bf  df e3 de bf 28 e2 a0 e0  |.$..........(...|
000a0  b8 e2 01 c0 1d 92 a4 30  b2 07 e1 f7 9b d0 26 c1  |.......0......&.|
000b0  a7 cf 90 91 01 0a 9f 7e  90 93 01 0a 90 91 60 04  |..... .~... ..`.|
000c0  93 ff 0a c0 81 11 04 c0  88 e0 80 93 66 04 08 95  |.. .........f...|
000d0  88 e0 80                                          |...             |

avrdude> read fuse0 0 0x10
>>> read fuse0 0 0x10 

Reading | ################################################## | 100% 0.00 s 

0000  00                                                |.               |

avrdude> read sig
>>> read sig 

Reading | ################################################## | 100% 0.00 s 

0000  1e 97 0c                                          |..              |

avrdude> read flash 0x1ffdc ...
>>> read flash 0x1ffdc ... 

Reading | ################################################## | 100% 0.00 s 

1ffdc  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
1ffec  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
1fffc  ff ff ff ff                                       |....            |

avrdude> quit
>>> quit 
avrdude> 
avrdude done.  Thank you.
```

</details>

<details>
<summary>Add a pad byte to the address if the memory that's being read from is greater than 64kiB</summary>

See how the number of padding bytes varies between flash and EEPROM read:

```
$ ./avrdude -pavr128db48 -cpkobn_updi -t -B0.5

         Vtarget                      : 3.31 V
         PDI/UPDI clock Xmega/megaAVR : 2000 kHz
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e970c (probably avr128db48)
avrdude> read flash 0 0x100
>>> read flash 0 0x100 

Reading | ################################################## | 100% 0.03 s 

00000  47 c0 00 00 55 c0 00 00  53 c0 00 00 51 c0 00 00  |G...U...S...Q...|
00010  4f c0 00 00 4d c0 00 00  4b c0 00 00 49 c0 00 00  |O...M...K...I...|
00020  47 c0 00 00 45 c0 00 00  43 c0 00 00 41 c0 00 00  |G...E...C...A...|
00030  3f c0 00 00 3d c0 00 00  3b c0 00 00 39 c0 00 00  |?...=...;...9...|
00040  37 c0 00 00 35 c0 00 00  33 c0 00 00 31 c0 00 00  |7...5...3...1...|
00050  2f c0 00 00 2d c0 00 00  2b c0 00 00 29 c0 00 00  |/...-...+...)...|
00060  27 c0 00 00 97 c0 00 00  23 c0 00 00 21 c0 00 00  |'.......#...!...|
00070  1f c0 00 00 1d c0 00 00  1b c0 00 00 19 c0 00 00  |................|
00080  17 c0 00 00 15 c0 00 00  13 c0 00 00 11 c0 00 00  |................|
00090  11 24 1f be cf ef cd bf  df e3 de bf 28 e2 a0 e0  |.$..........(...|
000a0  b8 e2 01 c0 1d 92 a4 30  b2 07 e1 f7 9b d0 26 c1  |.......0......&.|
000b0  a7 cf 90 91 01 0a 9f 7e  90 93 01 0a 90 91 60 04  |..... .~... ..`.|
000c0  93 ff 0a c0 81 11 04 c0  88 e0 80 93 66 04 08 95  |.. .........f...|
000d0  88 e0 80 93 65 04 08 95  9f b7 f8 94 81 11 07 c0  |....e...........|
000e0  80 91 73 04 87 7f 80 93  73 04 9f bf 08 95 80 91  |..s.....s.......|
000f0  73 04 88 60 f8 cf 8f b7  f8 94 20 91 00 28 30 91  |s..`...... ..(0.|

avrdude> read eeprom 0 0x80
>>> read eeprom 0 0x80 

Reading | ################################################## | 100% 0.52 s 

0000  48 65 6c 6c 6f 20 57 6f  72 6c 64 20 45 45 50 52  |Hello World EEPR|
0010  4f 4d 20 77 72 69 74 65  20 74 65 73 74 00 ff ff  |OM write test...|
0020  41 56 52 44 55 44 45 20  2d 20 41 56 52 20 44 6f  |AVRDUDE - AVR Do|
0030  77 6e 6c 6f 61 64 65 72  20 55 70 6c 6f 61 64 65  |wnloader Uploade|
0040  72 20 2d 20 69 73 20 61  20 70 72 6f 67 72 61 6d  |r - is a program|
0050  20 66 6f 72 20 64 6f 77  6e 6c 6f 61 64 69 6e 67  | for downloading|
0060  20 61 6e 64 20 75 70 6c  6f 61 64 69 6e 67 20 74  | and uploading t|
0070  68 65 20 6f 6e 2d 63 68  69 70 20 6d 65 6d 6f 72  |he on-chip memor|

avrdude> quit
>>> quit 
avrdude> 
avrdude done.  Thank you.
```

</details>


#### What I'd like to add to this PR if @dl8dtl and/or @stefanrueger agree:

* Remove the `>>>` echo. Why is this needed? It only takes up screen estate, and only prints exactly what you just typed, which is also printed on the screen, on the line above.
* Add a dedicated `>>>` output to terminal read to show the user what the **read** or **read [mem]** command is actually doing, (and print this if verbosity level is 1 or greater?)

<details>
<summary>Example output:</summary>

```
$ ./avrdude -pavr128db48 -cpkobn_updi -t -B0.5

         Vtarget                      : 3.31 V
         PDI/UPDI clock Xmega/megaAVR : 2000 kHz
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e970c (probably avr128db48)
avrdude> read flash 0 0x20
>>> read flash 0 0x20 

Reading | ################################################## | 100% 0.03 s 

00000  47 c0 00 00 55 c0 00 00  53 c0 00 00 51 c0 00 00  |G...U...S...Q...|
00010  4f c0 00 00 4d c0 00 00  4b c0 00 00 49 c0 00 00  |O...M...K...I...|

avrdude> read 
>>> read 

Reading | ################################################## | 100% 0.00 s 

00020  47 c0 00 00 45 c0 00 00  43 c0 00 00 41 c0 00 00  |G...E...C...A...|
00030  3f c0 00 00 3d c0 00 00  3b c0 00 00 39 c0 00 00  |?...=...;...9...|

avrdude> verbose 1
>>> verbose 1 
New verbosity level: 1
avrdude> read
>>> read 
>>> read flash 0x40 0x20
Reading | ################################################## | 100% 0.00 s 

00040  37 c0 00 00 35 c0 00 00  33 c0 00 00 31 c0 00 00  |7...5...3...1...|
00050  2f c0 00 00 2d c0 00 00  2b c0 00 00 29 c0 00 00  |/...-...+...)...|

avrdude> read
>>> read 
>>> read flash 0x60 0x20
Reading | ################################################## | 100% 0.00 s 

00060  27 c0 00 00 97 c0 00 00  23 c0 00 00 21 c0 00 00  |'.......#...!...|
00070  1f c0 00 00 1d c0 00 00  1b c0 00 00 19 c0 00 00  |................|

avrdude> read
>>> read 
>>> read flash 0x80 0x20
Reading | ################################################## | 100% 0.00 s 

00080  17 c0 00 00 15 c0 00 00  13 c0 00 00 11 c0 00 00  |................|
00090  11 24 1f be cf ef cd bf  df e3 de bf 28 e2 a0 e0  |.$..........(...|

avrdude> read eeprom
>>> read eeprom 
>>> read eeprom 0x0 0x100
Reading | ################################################## | 100% 1.03 s 

0000  48 65 6c 6c 6f 20 57 6f  72 6c 64 20 45 45 50 52  |Hello World EEPR|
0010  4f 4d 20 77 72 69 74 65  20 74 65 73 74 00 ff ff  |OM write test...|
0020  41 56 52 44 55 44 45 20  2d 20 41 56 52 20 44 6f  |AVRDUDE - AVR Do|
0030  77 6e 6c 6f 61 64 65 72  20 55 70 6c 6f 61 64 65  |wnloader Uploade|
0040  72 20 2d 20 69 73 20 61  20 70 72 6f 67 72 61 6d  |r - is a program|
0050  20 66 6f 72 20 64 6f 77  6e 6c 6f 61 64 69 6e 67  | for downloading|
0060  20 61 6e 64 20 75 70 6c  6f 61 64 69 6e 67 20 74  | and uploading t|
0070  68 65 20 6f 6e 2d 63 68  69 70 20 6d 65 6d 6f 72  |he on-chip memor|
0080  69 65 73 20 6f 66 20 4d  69 63 72 6f 63 68 69 70  |ies of Microchip|
0090  e2 80 99 73 20 41 56 52  20 6d 69 63 72 6f 63 6f  |...s AVR microco|
00a0  6e 74 72 6f 6c 6c 65 72  73 2e 20 49 74 20 63 61  |ntrollers. It ca|
00b0  6e 20 70 72 6f 67 72 61  6d 20 74 68 65 20 46 6c  |n program the Fl|
00c0  61 73 68 20 61 6e 64 20  45 45 50 52 4f 4d 2c 20  |ash and EEPROM, |
00d0  61 6e 64 20 77 68 65 72  65 20 73 75 70 70 6f 72  |and where suppor|
00e0  74 65 64 20 62 79 20 74  68 65 20 70 72 6f 67 72  |ted by the progr|
00f0  61 6d 6d 69 6e 67 20 70  72 6f 74 6f 63 6f 6c 2c  |amming protocol,|

avrdude> read eeprom
>>> read eeprom 
>>> read eeprom 0x100 0x100
Reading | ################################################## | 100% 1.03 s 

0100  20 69 74 20 63 61 6e 20  70 72 6f 67 72 61 6d 20  | it can program |
0110  66 75 73 65 20 61 6e 64  20 6c 6f 63 6b 20 62 69  |fuse and lock bi|
0120  74 73 2e 20 41 56 52 44  55 44 45 20 61 6c 73 6f  |ts. AVRDUDE also|
0130  20 73 75 70 70 6c 69 65  73 20 61 20 64 69 72 65  | supplies a dire|
0140  63 74 20 69 6e 73 74 72  75 63 74 69 6f 6e 20 6d  |ct instruction m|
0150  6f 64 65 20 61 6c 6c 6f  77 69 6e 67 20 6f 6e 65  |ode allowing one|
0160  20 74 6f 20 69 73 73 75  65 20 61 6e 79 20 70 72  | to issue any pr|
0170  6f 67 72 61 6d 6d 69 6e  67 20 69 6e 73 74 72 75  |ogramming instru|
0180  63 74 69 6f 6e 20 74 6f  20 74 68 65 20 41 56 52  |ction to the AVR|
0190  20 63 68 69 70 20 72 65  67 61 72 64 6c 65 73 73  | chip regardless|
01a0  20 6f 66 20 77 68 65 74  68 65 72 20 41 56 52 44  | of whether AVRD|
01b0  55 44 45 20 69 6d 70 6c  65 6d 65 6e 74 73 20 74  |UDE implements t|
01c0  68 61 74 20 73 70 65 63  69 66 69 63 20 66 65 61  |hat specific fea|
01d0  74 75 72 65 20 6f 66 20  61 20 70 61 72 74 69 63  |ture of a partic|
01e0  75 6c 61 72 20 63 68 69  70 2e 00 ff ff ff ff ff  |ular chip.......|
01f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> read
>>> read 
>>> read eeprom 0x200 0x100
Reading | ################################################## | 100% 0.00 s 

0000  48 65 6c 6c 6f 20 57 6f  72 6c 64 20 45 45 50 52  |Hello World EEPR|
0010  4f 4d 20 77 72 69 74 65  20 74 65 73 74 00 ff ff  |OM write test...|
0020  41 56 52 44 55 44 45 20  2d 20 41 56 52 20 44 6f  |AVRDUDE - AVR Do|
0030  77 6e 6c 6f 61 64 65 72  20 55 70 6c 6f 61 64 65  |wnloader Uploade|
0040  72 20 2d 20 69 73 20 61  20 70 72 6f 67 72 61 6d  |r - is a program|
0050  20 66 6f 72 20 64 6f 77  6e 6c 6f 61 64 69 6e 67  | for downloading|
0060  20 61 6e 64 20 75 70 6c  6f 61 64 69 6e 67 20 74  | and uploading t|
0070  68 65 20 6f 6e 2d 63 68  69 70 20 6d 65 6d 6f 72  |he on-chip memor|
0080  69 65 73 20 6f 66 20 4d  69 63 72 6f 63 68 69 70  |ies of Microchip|
0090  e2 80 99 73 20 41 56 52  20 6d 69 63 72 6f 63 6f  |...s AVR microco|
00a0  6e 74 72 6f 6c 6c 65 72  73 2e 20 49 74 20 63 61  |ntrollers. It ca|
00b0  6e 20 70 72 6f 67 72 61  6d 20 74 68 65 20 46 6c  |n program the Fl|
00c0  61 73 68 20 61 6e 64 20  45 45 50 52 4f 4d 2c 20  |ash and EEPROM, |
00d0  61 6e 64 20 77 68 65 72  65 20 73 75 70 70 6f 72  |and where suppor|
00e0  74 65 64 20 62 79 20 74  68 65 20 70 72 6f 67 72  |ted by the progr|
00f0  61 6d 6d 69 6e 67 20 70  72 6f 74 6f 63 6f 6c 2c  |amming protocol,|

avrdude> quit
>>> quit 
avrdude> 
avrdude done.  Thank you.
```

</details>

mcuee commented:
@MCUdude 

The verbose level does not seem to work as per your example.

```
C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1209 -pm4808 -cpickit4_updi -t

                Vtarget                      : 4.90 V
                JTAG clock megaAVR/program   : 1000 kHz
                JTAG clock megaAVR/debug     : 100 kHz
                PDI/UPDI clock Xmega/megaAVR : 100 kHz
avrdude_pr1209: AVR device initialized and ready to accept instructions
avrdude_pr1209: device signature = 0x1e9650 (probably m4808)
avrdude> read
>>> read
Usage: read <memory> <addr> <len>
       read <memory> <addr> ...
       read <memory> <addr>
       read <memory> ...
       read <memory>
avrdude> read flash 0 0x40
>>> read flash 0 0x40

Reading | ################################################## | 100% 0.21 s

0000  01 c0 d8 c0 11 24 80 91  40 00 88 23 69 f0 28 2f  |.....$..@..#i.(/|
0010  30 e0 83 fd 03 c0 2d 7f  23 2b 31 f4 80 93 40 00  |0.....-.#+1...@.|
0020  8c bb 80 e0 af d0 ec c0  a8 95 a0 9a a8 9a 10 92  |................|
0030  e2 05 80 91 82 12 83 70  81 30 29 f5 8c e5 90 e0  |.......p.0).....|

avrdude> verbose 1
>>> verbose 1
New verbosity level: 1
avrdude> read
>>> read

Reading | ################################################## | 100% 0.17 s

0040  80 93 48 08 90 93 49 08  81 e0 80 93 4b 08 83 e0  |..H...I.....K...|
0050  80 93 47 08 10 92 45 08  80 ec 80 93 46 08 8b e0  |..G...E.....F...|
0060  91 d0 42 9a 87 e0 81 50  89 f4 a8 95 0d e9 83 e0  |..B....P.... ...|
0070  d8 2e 7c d0 81 34 f1 f4  79 d0 18 2f 8c d0 12 38  |..|..4..y../...8|

avrdude> read
>>> read

Reading | ################################################## | 100% 0.20 s

0080  89 f4 80 e0 13 c0 83 e7  90 e0 da cf 52 9a 2e ec  |............R...|
0090  36 e5 a8 95 90 91 44 08  97 fd e8 cf 21 50 31 09  |6.....D.....!P1 |
00a0  c1 f7 e1 cf 89 e0 11 38  09 f0 83 e0 58 d0 80 e1  |.......8 ...X...|
00b0  56 d0 df cf 82 34 19 f4  84 e1 75 d0 f8 cf 85 34  |V....4....u....4|

avrdude> verbose 2
>>> verbose 2
New verbosity level: 2
avrdude> read
>>> read

Reading | ################################################## | 100% 0.18 s

00c0  11 f4 85 e0 fa cf 85 35  31 f4 50 d0 c8 2f 4e d0  |.......51.P../N.|
00d0  d8 2f 61 d0 ec cf 86 35  19 f4 84 e0 64 d0 d1 cf  |./a....5....d...|
00e0  84 36 b9 f4 43 d0 42 d0  18 2f 40 d0 86 34 79 f4  |.6..C.B../@..4y.|
00f0  d0 5c 3c d0 88 83 21 96  11 50 d9 f7 4c d0 04 bf  |.\<...!..P..L...|

avrdude> quit
>>> quit
avrdude> avrdude_pr1209: sending leave progmode command: 0x80 (3 bytes msg)
avrdude_pr1209: jtag3_close()
avrdude_pr1209: sending AVR sign-off command: 0x80 (3 bytes msg)
avrdude_pr1209: sending sign-off command: 0x80 (3 bytes msg)

avrdude_pr1209 done.  Thank you.

```

mcuee commented:
The rollover behavior seems to be consistent.
```
C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1209 -pm4808 -cpickit4_updi -qt

                Vtarget                      : 4.92 V
                JTAG clock megaAVR/program   : 1000 kHz
                JTAG clock megaAVR/debug     : 100 kHz
                PDI/UPDI clock Xmega/megaAVR : 100 kHz
avrdude_pr1209: AVR device initialized and ready to accept instructions
avrdude_pr1209: device signature = 0x1e9650 (probably m4808)
avrdude> dump fuses 3
>>> dump fuses 3
0003  ff 00 c4 07 00 02                                 |......          |
0000  00 54 02                                          |.T.             |

avrdude> dump fuses
>>> dump fuses
0000  00 54 02 ff 00 c4 07 00  02                       |.T.......       |

avrdude> dump flash 0xbfc0 18
>>> dump flash 0xbfc0 18
bfc0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
bfd0  ff ff                                             |..              |

avrdude> dump
>>> dump
bfd2  42 6c 69 6e 6b 2e 69 6e  6f 5f 61 74 6d 65 67 61  |Blink.ino_atmega|
bfe2  34 38                                             |48              |

avrdude> dump
>>> dump
bfe4  30 38 5f 31 36 30 30 30  30 30 30 4c 2e 68 65 78  |08_16000000L.hex|
bff4  00 e6                                             |..              |

avrdude> dump
>>> dump
bff6  07 0b 1b 13 1e f6 04 dc  ba 23                    |. .......#      |
0000  01 c0 d8 c0 11 24 80 91                           |.....$..        |

avrdude> dump
>>> dump
0000  01 c0 d8 c0 11 24 80 91  40 00 88 23 69 f0 28 2f  |.....$..@..#i.(/|
0010  30 e0                                             |0.              |

avrdude> quit
>>> quit
avrdude>
avrdude_pr1209 done.  Thank you.
```

MCUdude commented:
> The verbose level does not seem to work as per your example. [Maybe this has not been implemented].

Correct. I have working code that I've not committed yet. I'd like to get @dl8dtl's permission before I remove the `>>>` echo and add a custom one for terminal read

stefanrueger commented:
> Remove the >>> echo [...] dedicated >>> output to terminal [when verbosity > 0]

@MCUdude Great idea. Thinking of it more I can see no use of the ">>>" echo, even when `HAVE_LIBREADLINE` is not defined (I believe that would only be useful if the user had deliberately switched off OS echo before) . Not sure whether @dl8dtl follows the details of terminal I/O, so I think you can assume it's OK unless @dl8dtl  intervenes or @mcuee finds an OS/library combination where the user would not see what they typed otherwise.

mcuee commented:
I am okay with the proposal to remove the `>>>` echo as well. And of course I will carry out the test once the PR is updated to have that feature.

MCUdude commented:
> I am okay with the proposal to remove the >>> echo as well. And of course I will carry out the test once the PR is updated to have that feature.

Great! I'll push the change when I'm home. 

> Not sure whether @dl8dtl follows the details of terminal I/O, so I think you can assume it's OK unless @dl8dtl intervenes or @mcuee finds an OS/library combination where the user would not see what they typed otherwise.

@mcuee Do you know if there exist any OS/shell combinations where the `>>>` is necessary because the command the user typed isn't printed? If not, I see no reason why we shouldn't remove the `>>>` echo

mcuee commented:
> @mcuee Do you know if there exist any OS/shell combinations where the `>>>` is necessary because the command the user typed isn't printed? If not, I see no reason why we shouldn't remove the `>>>` echo

I have not encunterered that situation myself with the recent tests across different Operating systems.

As for shell, I believe most of the Unix shell should perform similar. I usually use default shell under Linux (bash or dash), macOS (zsh, occassionally bash) and BSDs (sh, ksh, csh). 

For Windows I tested with Power Shell (`-B0.5` does not work but `-B 0.5` works), MSYS2 shell and Windows command prompt. 

So far so good.

mcuee commented:
@MCUdude 

It seems to work well, Only minor thing is to have a space after the length in the echo. 
https://github.com/avrdudes/avrdude/pull/1209/commits/ffa6d32bb21f104b2ee94f62548915923472e8a5

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1209v1 -c usbasp -p m328p -qt

avrdude_pr1209v1: AVR device initialized and ready to accept instructions
avrdude_pr1209v1: device signature = 0x1e950f (probably m328p)
avrdude> dump lfuse
0000  f7                                                |.               |

avrdude> dump
0000  f7                                                |.               |

avrdude> dump eeprom 0 0x10
0000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> dump
0010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> verbose
Verbosity level: 0
avrdude> verbose 1
New verbosity level: 1
avrdude> dump
>>> dump eeprom 0x20 0x100020  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> dump flash 0 0x10
0000  0c 94 b5 00 0c 94 cd 06  0c 94 d2 00 0c 94 d2 00  | ... ... ... ...|

avrdude> dump
>>> dump flash 0x10 0x100010  0c 94 d2 00 0c 94 d2 00  0c 94 d2 00 0c 94 d2 00  | ... ... ... ...|
                                   (should be 0x10 0010)
avrdude> verbose
Verbosity level: 1
avrdude> dump
>>> dump flash 0x20 0x100020  0c 94 d2 00 0c 94 d2 00  0c 94 d2 00 0c 94 d2 00  | ... ... ... ...|
                                  (should be 0x10 0020)
avrdude> quit
avrdude>
avrdude_pr1209v1 done.  Thank you.
```

mcuee commented:
I have figured out a hacky way to build avrdude using MSVC and GNU Readline.

This PR seems to work fine.
```
C:\work\avr\avrdude_test\avrdude_bin\avrdude_msvc_readline_pr1209> .\avrdude -c xplainedpro_updi -p t817 -qqt
avrdude> dump flash 0 0x40
0000  01 c0 d8 c0 11 24 80 91  40 00 88 23 69 f0 28 2f  |.....$..@..#i.(/|
0010  30 e0 83 fd 03 c0 2d 7f  23 2b 31 f4 80 93 40 00  |0.....-.#+1...@.|
0020  8c bb 80 e0 af d0 ec c0  a8 95 22 9a 2a 9a 10 92  |..........".*...|
0030  01 02 80 91 82 12 83 70  81 30 61 f5 8c e5 90 e0  |.......p.0a.....|

avrdude> dump
0040  80 93 08 08 90 93 09 08  81 e0 80 93 0b 08 83 e0  |...... ..... ...|
0050  80 93 07 08 10 92 05 08  80 ec 80 93 06 08 8b e0  |................|
0060  91 d0 07 9a 87 e0 81 50  c1 f4 a8 95 0d e9 83 e0  |.......P.... ...|
0070  d8 2e 7c d0 81 34 e9 f4  79 d0 18 2f 8c d0 81 e0  |..|..4..y../....|

avrdude> dump fuses
0000  00 00 02 ff 00 c4 04 00  02                       |.........       |

avrdude> dump
0000  00 00 02 ff 00 c4 04 00  02                       |.........       |

avrdude> dump flash
0080  12 38 21 f0 89 e0 11 38  09 f0 83 e0 68 d0 80 e1  |.8!....8 ...h...|
0090  66 d0 ef cf 83 e7 90 e0  d3 cf 17 9a 2e ec 36 e5  |f.............6.|
00a0  a8 95 90 91 04 08 97 fd  e1 cf 21 50 31 09 c1 f7  |..........!P1 ..|
00b0  da cf 82 34 19 f4 84 e1  76 d0 e9 cf 85 34 11 f4  |...4....v....4..|

avrdude> verbose 1
New verbosity level: 1
avrdude> dump flash
>>> dump flash 0xc0 0x4000c0  85 e0 fa cf 85 35 31 f4  51 d0 c8 2f 4f d0 d8 2f  |.....51.Q../O../| 
(missing a space and return here, same as the previous comment)
00d0  62 d0 dd cf 86 35 21 f4  84 e0 65 d0 80 e0 d6 cf  |b....5!...e.....|
00e0  84 36 b9 f4 43 d0 42 d0  18 2f 40 d0 86 34 79 f4  |.6..C.B../@..4y.|
00f0  d0 58 3c d0 88 83 21 96  11 50 d9 f7 4c d0 04 bf  |.X<...!..P..L...|

avrdude> quit
avrdude>
avrdude done.  Thank you.

```

mcuee commented:
Same for default MSVC build (without GNU Readline) -- I just use the binary from github action.

```
C:\work\avr\avrdude_test\avrdude_bin\avrdude-msvc_pr1209> .\avrdude -c xplainedpro_updi -p t817 -qqt
avrdude> dump
Usage: dump <memory> <addr> <len>
       dump <memory> <addr> ...
       dump <memory> <addr>
       dump <memory> ...
       dump <memory>
avrdude> dump flash 0 0x20
0000  01 c0 d8 c0 11 24 80 91  40 00 88 23 69 f0 28 2f  |.....$..@..#i.(/|
0010  30 e0 83 fd 03 c0 2d 7f  23 2b 31 f4 80 93 40 00  |0.....-.#+1...@.|

avrdude> dump
0020  8c bb 80 e0 af d0 ec c0  a8 95 22 9a 2a 9a 10 92  |..........".*...|
0030  01 02 80 91 82 12 83 70  81 30 61 f5 8c e5 90 e0  |.......p.0a.....|

avrdude> verbose 1
New verbosity level: 1
avrdude> dump
>>> dump flash 0x40 0x200040  80 93 08 08 90 93 09 08  81 e0 80 93 0b 08 83 e0  |...... ..... ...|
0050  80 93 07 08 10 92 05 08  80 ec 80 93 06 08 8b e0  |................|

avrdude> dump
>>> dump flash 0x60 0x200060  91 d0 07 9a 87 e0 81 50  c1 f4 a8 95 0d e9 83 e0  |.......P.... ...|
0070  d8 2e 7c d0 81 34 e9 f4  79 d0 18 2f 8c d0 81 e0  |..|..4..y../....|

avrdude> verbose 0
New verbosity level: 0
avrdude> dump fuses 0x2
0002  02 ff 00 c4 04 00 02                              |.......         |
0000  00 00                                             |..              |

avrdude> dump
0002  02 ff 00 c4 04 00 02                              |.......         |
0000  00 00                                             |..              |

avrdude> dump fuses 0
0000  00 00 02 ff 00 c4 04 00  02                       |.........       |

avrdude> quit

```

mcuee commented:
> missing a return in the echo (verbose mode)

This is now fixed and it is pretty good now.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1209v2 -c urclock -P COM1 -p m328p -xbootsize=512 -qqt
avrdude> dump
Usage: dump <memory> <addr> <len>
       dump <memory> <addr> ...
       dump <memory> <addr>
       dump <memory> ...
       dump <memory>
avrdude> dump flash 0 0x10
0000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> verbose 1
New verbosity level: 1
avrdude> dump
>>> dump flash 0x10 0x10
0010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

avrdude> quit
avrdude>
avrdude_pr1209v2 done.  Thank you.

```

mcuee commented:
@MCUdude and @stefanrueger 

I am thinking that there is a missing return somewhere (at least for Linux) when using `-qq -t`. This is not the issue of this PR but rather already in git main.

Not a big issue but I am not so sure if there is an easy fix.

```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1209v2 -c urclock -P /dev/ttyACM0 -p m328p -qqt
avrdude> quit
avrdude> mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ 
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1209v2 -c urclock -P /dev/ttyACM0 -p m328p -qqt
avrdude> quit
avrdude> mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$  (I have to hit return here to feel better).
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1209v2 -c urclock -P /dev/ttyACM0 -p m328p -qt

avrdude_pr1209v2: AVR device initialized and ready to accept instructions
avrdude_pr1209v2: device signature = 0x1e950f (probably m328p)
avrdude> quit
avrdude> 
avrdude_pr1209v2 done.  Thank you.

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ 
```

mcuee commented:
I have also done some simple tests under macOS and it works fine there as well (with either libedit or GNU readline).

I think this PR can be merged now.

stefanrueger commented:
> missing return somewhere (at least for Linux) when using `-qq -t` 

@mcuee Interesting. You are right. I never noticed (seems to work for me under ubuntu with GNU libredline), but I agree there is an extra `avrdude>` prompt, and it can be *mildly* annoying: it also crops up when redirecting output to a file, but strangely I do not see it interactively (and my prompt does not appear to start with a carriage return either...):
```
srueger@medlar $ avrdude -qqt
avrdude> quit
srueger@medlar $ avrdude -qqt >/tmp/1 </dev/null
srueger@medlar $ cat /tmp/1
avrdude> 
avrdude> srueger@medlar $
```
I have not the faintest how to solve this; ideas on a postcard, please ...

stefanrueger reviewed (commented):
There is subtle problem [here:](https://github.com/MCUdude/avrdude/blob/41603c7b153a3d15f1e0ad8fb8afa4c9d80ded71/src/term.c#L309) `strtol()` can trigger a RANGE error that needs to be checked by setting `errno = 0` before the call and checking for `errno != 0` afterwards. It is much easier to use `strtoul()` (note the **u**) which cannot have a range error (unsigned in C relies on modulo arithmetic that cannot overflow). The code then correctly checks against < 0 (unsigned comparison against MAX_INT); the error message is not wrong, but mildly confusing when you type `read flash 0 -1`. However, nicer perhaps to default the length to 1 in this case, the smallest possible length?

Also note a memory leak might ensue when the allocated buf is not freed before returning. Here the summary of my proposed changes:
```
diff --git a/src/term.c b/src/term.c
index 99eff97..0792bde 100644
--- a/src/term.c
+++ b/src/term.c
@@ -306,16 +306,13 @@ static int cmd_dump(PROGRAMMER *pgm, AVRPART *p, int argc, char *argv[]) {
         read_mem[i].addr = 0;
       read_mem[i].len = maxsize - read_mem[i].addr;
     } else if (argc == 4) {
-      unsigned long ul = strtol(argv[3], &end_ptr, 0);
+      unsigned long ul = strtoul(argv[3], &end_ptr, 0);
       if (*end_ptr || (end_ptr == argv[3])) {
         pmsg_error("(dump) cannot parse length %s\n", argv[3]);
         return -1;
       }
-      if (ul > INT_MAX) {
-        pmsg_error("(dump) %s cannot read 0x%lx bytes from start address 0x%0*x\n", mem->desc, ul,
-          mem->size > 0x10000? 5: 4, maxsize-1);
-        return -1;
-      }
+      if (ul == 0 || ul > INT_MAX) // Not positive if used as int, make it 1
+        ul = 1;
       read_mem[i].len = (int) ul;
     }
   }
@@ -345,6 +342,7 @@ static int cmd_dump(PROGRAMMER *pgm, AVRPART *p, int argc, char *argv[]) {
       pmsg_error("(dump) error reading %s address 0x%05lx of part %s\n", mem->desc, (long) read_mem[i].addr + j, p->desc);
       if (rc == -1)
         imsg_error("%*sread operation not supported on memory type %s\n", 7, "", mem->desc);
+      free(buf);
       return -1;
     }
     report_progress(j, read_mem[i].len, NULL);
```

I can push the changes above to this PR, or @MCUdude can check and change...

mcuee commented:
> > missing return somewhere (at least for Linux) when using `-qq -t`
> 
> @mcuee Interesting. You are right. I never noticed (seems to work for me under ubuntu with GNU libredline), but I agree there is an extra `avrdude>` prompt, and it can be _mildly_ annoying: it also crops up when redirecting output to a file, but strangely I do not see it interactively (and my prompt does not appear to start with a carriage return either...):
> 
> ```
> srueger@medlar $ avrdude -qqt
> avrdude> quit
> srueger@medlar $ avrdude -qqt >/tmp/1 </dev/null
> srueger@medlar $ cat /tmp/1
> avrdude> 
> avrdude> srueger@medlar $
> ```
> 
> I have not the faintest how to solve this; ideas on a postcard, please ...

I will create a seperate issue for this one. It is not a big problem though.

dl8dtl commented:
> #### What I'd like to add to this PR if @dl8dtl and/or @stefanrueger agree:
> 
>     * Remove the `>>>` echo. Why is this needed?

I think it has "always" been there, only later it got conditionalized (which I don't see the reason for).
I think it can just go away.

MCUdude commented:
> There is subtle problem [here:](https://github.com/MCUdude/avrdude/blob/41603c7b153a3d15f1e0ad8fb8afa4c9d80ded71/src/term.c#L309) strtol() can trigger a RANGE error that needs to be checked by setting errno = 0 before the call and checking for errno != 0 afterwards. It is much easier to use strtoul() (note the u) which cannot have a range error (unsigned in C relies on modulo arithmetic that cannot overflow). The code then correctly checks against < 0 (unsigned comparison against MAX_INT); the error message is not wrong, but mildly confusing when you type read flash 0 -1. However, nicer perhaps to default the length to 1 in this case, the smallest possible length? Also note a memory leak might ensue when the allocated buf is not freed before returning. 

Thanks, @stefanrueger! The remaining `strtol` was a mistake from my side. I've implemented your other suggestions as well.

> I think it has "always" been there, only later it got conditionalized (which I don't see the reason for).
I think it can just go away.

I'm glad you agree, as I've already removed it 😉

stefanrueger commented:
I'll merge this in the next few days. I tend to prefer a squash & merge for PRs that test a few things before converging. Any objections for that?
